agent/http_decode_test.go

```diff
@@ -0,0 +1,3304 @@
+//   331  	}
+// ==================================
+
+// ACLPolicySetRequest:
+// Policy	structs.ACLPolicy
+//     ID	string
+//     Name	string
+//     Description	string
+//     Rules	string
+//     Syntax	acl.SyntaxVersion
+//     Datacenters	[]string
+//     Hash	[]uint8
+//     RaftIndex	structs.RaftIndex
+//         CreateIndex	uint64
+//         ModifyIndex	uint64
+// Datacenter	string
+// WriteRequest	structs.WriteRequest
+//     Token	string
+func TestDecodeACLPolicyWrite(t *testing.T) {
+
+	for _, tc := range hashTestCases {
+		t.Run(tc.desc, func(t *testing.T) {
+			jsonStr := fmt.Sprintf(`{
+		"Hash": %s
+	}`, tc.hashes.in)
+			body := bytes.NewBuffer([]byte(jsonStr))
+			req := httptest.NewRequest("POST", "http://foo.com", body)
+
+			var out structs.ACLPolicy
+			err := decodeBody(req, &out, fixTimeAndHashFields)
+			if err != nil && !tc.wantErr {
+				t.Fatal(err)
+			}
+			if err == nil && tc.wantErr {
+				t.Fatal("expected error, got nil")
+			}
+			if !bytes.Equal(out.Hash, tc.hashes.want) {
+				t.Fatalf("expected hash to be %s, got %s", tc.hashes.want, out.Hash)
+			}
+		})
+
+	}
+}
+
+// ==================================
+// $GOPATH/github.com/hashicorp/consul/agent/acl_endpoint.go:
+//   511  	s.parseToken(req, &args.Token)
+//   512
+//   513: 	if err := decodeBody(req, &args.ACLToken, fixTimeAndHashFields); err != nil {
+//   514  		return nil, BadRequestError{Reason: fmt.Sprintf("Token decoding failed: %v", err)}
+//   515  	}
+// ==================================
+
+// ACLTokenSetRequest:
+// ACLToken	structs.ACLToken
+//     AccessorID	string
+//     SecretID	string
+//     Description	string
+//     Policies	[]structs.ACLTokenPolicyLink
+//         ID	string
+//         Name	string
+//     Roles	[]structs.ACLTokenRoleLink
+//         ID	string
+//         Name	string
+//     ServiceIdentities	[]*structs.ACLServiceIdentity
+//         ServiceName	string
+//         Datacenters	[]string
+//     Type	string
+//     Rules	string
+//     Local	bool
+//     AuthMethod	string
+//     ExpirationTime	*time.Time
+//     ExpirationTTL	time.Duration
+//     CreateTime	time.Time
+//     Hash	[]uint8
+//     RaftIndex	structs.RaftIndex
+//         CreateIndex	uint64
+//         ModifyIndex	uint64
+// Create	bool
+// Datacenter	string
+// WriteRequest	structs.WriteRequest
+//     Token	string
+
+func TestDecodeACLToken(t *testing.T) {
+
+	for _, tc := range translateValueTestCases {
+		t.Run(tc.desc, func(t *testing.T) {
+			// set up request body
+			var expTime, expTTL, createTime, hash = "null", "null", "null", "null"
+			if tc.hashes != nil {
+				hash = tc.hashes.in
+			}
+			if tc.timestamps != nil {
+				expTime = tc.timestamps.in
+				createTime = tc.timestamps.in
+			}
+			if tc.durations != nil {
+				expTTL = tc.durations.in
+			}
+			bodyBytes := []byte(fmt.Sprintf(`{
+		"ExpirationTime": %s,
+		"ExpirationTTL": %s,
+		"CreateTime": %s,
+		"Hash": %s
+	}`, expTime, expTTL, createTime, hash))
```

Comment:
Given that the indentation in the body shouldn't matter to any JSON parser, is there a reason not to indent this a little more so the code is easier to read?

Not a big deal if you don't agree it's better, but I found this a little confusing to scan at first and thought this was the end of `t.Run()` call.

Reply:
totally, no you're right. i'll run through all the other tests too to make sure they're indented OK.